File: deprecated/server/src/chat/routes/workspaces.ts
/**
 * workspaces.ts — REST endpoints for workspace CRUD + file management
 *
 * Mirrors the Python backend's /workspaces endpoints.
 * Workspaces live in `workspaces/{id}/config.json` + `workspaces/{id}/files/`.
 *
 * Prefix: /chat/api  (applied externally when mounting)
 */

import { Router } from "express";
import * as fs from "node:fs/promises";
import * as fsSync from "node:fs";
import * as path from "node:path";

// ── Helpers ─────────────────────────────────────────────────────────────────

function sanitizeId(name: string): string {
  return (
    name
      .toLowerCase()
      .replace(/[^a-z0-9_-]/g, "-")
      .replace(/^-+|-+$/g, "") || "workspace"
  );
}

// ── Router ──────────────────────────────────────────────────────────────────

export function createWorkspaceRoutes(projectRoot: string): Router {
  const router = Router();
  const workspacesDir = path.join(projectRoot, "workspaces");
  const historyDir = path.join(projectRoot, "history");

  // Ensure directory exists
  fsSync.mkdirSync(workspacesDir, { recursive: true });

  // ── GET /workspaces ───────────────────────────────────────────────────────
  router.get("/workspaces", async (_req, res) => {
    try {
      let entries: string[];
      try {
        entries = await fs.readdir(workspacesDir);
      } catch {
        res.json([]);
        return;
      }

      entries.sort();
      const result: Array<Record<string, unknown>> = [];

      for (const entry of entries) {
        const dir = path.join(workspacesDir, entry);
        const stat = await fs.stat(dir).catch(() => null);
        if (!stat?.isDirectory()) continue;

        const configPath = path.join(dir, "config.json");
        try {
          const raw = await fs.readFile(configPath, "utf-8");
          result.push(JSON.parse(raw) as Record<string, unknown>);
        } catch {
          // skip malformed
        }
      }

      res.json(result);
    } catch (err: any) {
      res.status(500).json({ error: err.message ?? "Internal error" });
    }
  });

  // ── POST /workspaces ──────────────────────────────────────────────────────
  router.post("/workspaces", async (req, res) => {
    try {
      const body = req.body as Record<string, unknown>;
      const name = ((body.name as string) ?? "").trim();
      if (!name) {
        res.status(400).json({ error: "name required" });
        return;
      }

      let workspaceId = sanitizeId(name);
      const base = workspaceId;
      let idx = 2;
      while (fsSync.existsSync(path.join(workspacesDir, workspaceId))) {
        workspaceId = `${base}-${idx}`;
        idx++;
      }

      const dir = path.join(workspacesDir, workspaceId);
      await fs.mkdir(dir, { recursive: true });
      await fs.mkdir(path.join(dir, "files"), { recursive: true });

      const cfg = {
        id: workspaceId,
        name,
        description: (body.description as string) ?? "",
        system_prompt: (body.system_prompt as string) ?? "",
        default_agents: (body.default_agents as string[]) ?? [],
        created_at: new Date().toISOString(),
      };

      await fs.writeFile(
        path.join(dir, "config.json"),
        JSON.stringify(cfg, null, 2),
        "utf-8",
      );
      res.json(cfg);
    } catch (err: any) {
      res.status(500).json({ error: err.message ?? "Internal error" });
    }
  });

  // ── GET /workspaces/:id ───────────────────────────────────────────────────
  router.get("/workspaces/:id", async (req, res) => {
    try {
      const workspaceId = req.params.id;
      const configPath = path.join(
        workspacesDir,
        workspaceId,
        "config.json",
      );

      let cfg: Record<string, unknown>;
      try {
        const raw = await fs.readFile(configPath, "utf-8");
        cfg = JSON.parse(raw) as Record<string, unknown>;
      } catch {
        res.status(404).json({ error: "Workspace not found" });
        return;
      }

      // List files
      const filesDir = path.join(workspacesDir, workspaceId, "files");
      let files: string[] = [];
      try {
        const entries = await fs.readdir(filesDir);
        files = entries
          .filter((f) => {
            try {
              return fsSync.statSync(path.join(filesDir, f)).isFile();
            } catch {
              return false;
            }
          })
          .sort();
      } catch {
        // files dir may not exist
      }

      res.json({ ...cfg, files });
    } catch (err: any) {
      res.status(500).json({ error: err.message ?? "Internal error" });
    }
  });

  // ── PUT /workspaces/:id ──────────────────────────────────────────────────
  router.put("/workspaces/:id", async (req, res) => {
    try {
      const workspaceId = req.params.id;
      const configPath = path.join(
        workspacesDir,
        workspaceId,
        "config.json",
      );

      let cfg: Record<string, unknown>;
      try {
        const raw = await fs.readFile(configPath, "utf-8");
        cfg = JSON.parse(raw) as Record<string, unknown>;
      } catch {
        res.status(404).json({ error: "Workspace not found" });
        return;
      }

      const body = req.body as Record<string, unknown>;
      for (const key of [
        "name",
        "description",
        "system_prompt",
        "default_agents",
      ]) {
        if (key in body) cfg[key] = body[key];
      }

      await fs.writeFile(
        configPath,
        JSON.stringify(cfg, null, 2),
        "utf-8",
      );
      res.json({ ok: true });
    } catch (err: any) {
      res.status(500).json({ error: err.message ?? "Internal error" });
    }
  });

  // ── DELETE /workspaces/:id ───────────────────────────────────────────────
  router.delete("/workspaces/:id", async (req, res) => {
    try {
      const workspaceId = req.params.id;
      const dir = path.join(workspacesDir, workspaceId);

      try {
        await fs.access(dir);
      } catch {
        res.status(404).json({ error: "Workspace not found" });
        return;
      }

      // Hard delete
      await fs.rm(dir, { recursive: true, force: true });

      // Detach sessions that belonged to this workspace
      try {
        const sessionEntries = await fs.readdir(historyDir);
        for (const sid of sessionEntries) {
          const mf = path.join(historyDir, sid, "messages.json");
          try {
            const raw = await fs.readFile(mf, "utf-8");
            const msgs = JSON.parse(raw) as Array<Record<string, unknown>>;
            let changed = false;
            for (const m of msgs) {
              if (m.workspace_id === workspaceId) {
                m.workspace_id = null;
                changed = true;
              }
            }
            if (changed) {
              await fs.writeFile(
                mf,
                JSON.stringify(msgs, null, 2),
                "utf-8",
              );
            }
          } catch {
            // skip
          }
        }
      } catch {
        // history dir may not exist
      }

      res.json({ ok: true });
    } catch (err: any) {
      res.status(500).json({ error: err.message ?? "Internal error" });
    }
  });

  // ── POST /workspaces/:id/files — File upload ─────────────────────────────
  // Note: expects raw body with filename in query or Content-Disposition.
  // For multipart, consider adding multer middleware externally.
  router.post("/workspaces/:id/files", async (req, res) => {
    try {
      const workspaceId = req.params.id;
      const filesDir = path.join(workspacesDir, workspaceId, "files");

      try {
        await fs.access(filesDir);
      } catch {
        res.status(404).json({ error: "Workspace not found" });
        return;
      }

      // Accept filename from query param or body field
      const body = req.body as Record<string, unknown>;
      const filename = ((req.query.filename as string) ??
        (body.filename as string) ??
        ""
      ).trim();
      if (!filename || filename.includes("/") || filename.includes("\\")) {
        res.status(400).json({ error: "Valid filename required" });
        return;
      }

      // If body has content (base64 or text), write it
      const content = (body.content as string) ?? "";
      const encoding = (body.encoding as string) ?? "utf-8";

      if (encoding === "base64") {
        await fs.writeFile(
          path.join(filesDir, filename),
          Buffer.from(content, "base64"),
        );
      } else {
        await fs.writeFile(
          path.join(filesDir, filename),
          content,
          "utf-8",
        );
      }

      res.json({ ok: true, filename });
    } catch (err: any) {
      res.status(500).json({ error: err.message ?? "Internal error" });
    }
  });

  // ── DELETE /workspaces/:id/files/:filename ───────────────────────────────
  router.delete("/workspaces/:id/files/:filename", async (req, res) => {
    try {
      const workspaceId = req.params.id;
      const filename = req.params.filename;
      const filePath = path.join(
        workspacesDir,
        workspaceId,
        "files",
        filename,
      );

      try {
        await fs.access(filePath);
      } catch {
        res.status(404).json({ error: "File not found" });
        return;
      }

      await fs.unlink(filePath);
      res.json({ ok: true });
    } catch (err: any) {
      res.status(500).json({ error: err.message ?? "Internal error" });
    }
  });

  return router;
}

File: routes/workspaces.py
"""Workspace CRUD and file upload endpoints."""
import json
import re
import shutil
from datetime import datetime

from fastapi import APIRouter, File, HTTPException, UploadFile

from core.security import validate_filename, safe_workspace_path

router = APIRouter()


def _workspace_config_path(workspace_id: str):
    import app as _app
    return _app.WORKSPACES_DIR / workspace_id / "config.json"


def _load_workspace_config(workspace_id: str) -> dict:
    p = _workspace_config_path(workspace_id)
    if not p.exists():
        raise HTTPException(status_code=404, detail="Workspace not found")
    return json.loads(p.read_text())


@router.get("/workspaces")
async def list_workspaces():
    import app as _app
    result = []
    for d in sorted(_app.WORKSPACES_DIR.iterdir(), key=lambda x: x.name):
        if not d.is_dir():
            continue
        cp = d / "config.json"
        if not cp.exists():
            continue
        try:
            cfg = json.loads(cp.read_text())
            result.append(cfg)
        except Exception:
            pass
    return result


@router.post("/workspaces")
async def create_workspace(body: dict):
    import app as _app
    name = (body.get("name") or "").strip()
    if not name:
        raise HTTPException(status_code=400, detail="name required")
    workspace_id = re.sub(r"[^a-z0-9_-]", "-", name.lower()).strip("-") or "workspace"
    base = workspace_id
    idx = 2
    while (_app.WORKSPACES_DIR / workspace_id).exists():
        workspace_id = f"{base}-{idx}"
        idx += 1
    d = _app.WORKSPACES_DIR / workspace_id
    d.mkdir(parents=True)
    (d / "files").mkdir()
    cfg = {
        "id": workspace_id,
        "name": name,
        "description": body.get("description", ""),
        "system_prompt": body.get("system_prompt", ""),
        "default_agents": body.get("default_agents", []),
        "created_at": datetime.now().isoformat(),
    }
    (d / "config.json").write_text(json.dumps(cfg, indent=2, ensure_ascii=False))
    return cfg


@router.get("/workspaces/{workspace_id}")
async def get_workspace(workspace_id: str):
    import app as _app
    cfg = _load_workspace_config(workspace_id)
    files_dir = _app.WORKSPACES_DIR / workspace_id / "files"
    files = [f.name for f in files_dir.iterdir() if f.is_file()] if files_dir.exists() else []
    return {**cfg, "files": sorted(files)}


@router.put("/workspaces/{workspace_id}")
async def update_workspace(workspace_id: str, body: dict):
    cfg = _load_workspace_config(workspace_id)
    for key in ("name", "description", "system_prompt", "default_agents"):
        if key in body:
            cfg[key] = body[key]
    _workspace_config_path(workspace_id).write_text(json.dumps(cfg, indent=2, ensure_ascii=False))
    return {"ok": True}


@router.delete("/workspaces/{workspace_id}")
async def delete_workspace(workspace_id: str):
    import app as _app
    d = _app.WORKSPACES_DIR / workspace_id
    if not d.exists():
        raise HTTPException(status_code=404, detail="Workspace not found")
    shutil.rmtree(d)
    # Detach sessions that belonged to this workspace
    for session_dir in _app.HISTORY_DIR.iterdir():
        if not session_dir.is_dir():
            continue
        mf = session_dir / "messages.json"
        if not mf.exists():
            continue
        try:
            msgs = json.loads(mf.read_text())
            changed = False
            for m in msgs:
                if m.get("workspace_id") == workspace_id:
                    m["workspace_id"] = None
                    changed = True
            if changed:
                mf.write_text(json.dumps(msgs, indent=2, ensure_ascii=False))
        except Exception:
            pass
    return {"ok": True}


@router.post("/workspaces/{workspace_id}/files")
async def upload_workspace_file(workspace_id: str, file: UploadFile = File(...)):
    import app as _app
    d = _app.WORKSPACES_DIR / workspace_id / "files"
    if not d.exists():
        raise HTTPException(status_code=404, detail="Workspace not found")
    try:
        validate_filename(file.filename)
        safe_workspace_path(str(d), file.filename)
    except ValueError as e:
        raise HTTPException(status_code=403, detail=str(e))
    content = await file.read()
    (d / file.filename).write_bytes(content)
    return {"ok": True, "filename": file.filename}


@router.delete("/workspaces/{workspace_id}/files/{filename}")
async def delete_workspace_file(workspace_id: str, filename: str):
    import app as _app
    p = _app.WORKSPACES_DIR / workspace_id / "files" / filename
    if not p.exists():
        raise HTTPException(status_code=404, detail="File not found")
    p.unlink()
    return {"ok": True}


@router.put("/sessions/{session_id}/workspace")
async def move_session_to_workspace(session_id: str, body: dict):
    """Change the workspace_id field in a session's messages."""
    import app as _app
    mf = _app.session_messages_path(session_id)
    if not mf.exists():
        raise HTTPException(status_code=404, detail="Session not found")
    workspace_id = body.get("workspace_id")  # None to detach
    if workspace_id and not (_app.WORKSPACES_DIR / workspace_id).exists():
        raise HTTPException(status_code=404, detail="Workspace not found")
    msgs = json.loads(mf.read_text())
    for m in msgs:
        m["workspace_id"] = workspace_id
    mf.write_text(json.dumps(msgs, indent=2, ensure_ascii=False))
    return {"ok": True}
